oximeter/instruments/src/kstat/virtual_machine.rs

```diff
@@ -0,0 +1,459 @@
+    fn name(&self) -> &'static str {
+        "virtual_machine"
+    }
+
+    fn field_names(&self) -> &'static [&'static str] {
+        &[
+            "silo_id",
+            "silo_name",
+            "project_id",
+            "project_name",
+            "instance_id",
+            "instance_name",
+        ]
+    }
+
+    fn field_types(&self) -> Vec<FieldType> {
+        vec![
+            FieldType::Uuid,
+            FieldType::String,
+            FieldType::Uuid,
+            FieldType::String,
+            FieldType::Uuid,
+            FieldType::String,
+        ]
+    }
+
+    fn field_values(&self) -> Vec<FieldValue> {
+        vec![
+            self.silo_id.into(),
+            self.silo_name.clone().into(),
+            self.project_id.into(),
+            self.project_name.clone().into(),
+            self.instance_id.into(),
+            self.instance_name.clone().into(),
+        ]
+    }
+}
+
+/// Metric tracking vCPU usage by state.
+#[derive(Clone, Debug, Metric)]
+pub struct VcpuUsage {
+    /// The vCPU ID.
+    pub vcpu_id: u32,
+    /// The state of the vCPU.
+    pub state: String,
+    /// The cumulative time spent in this state, in nanoseconds.
+    pub datum: Cumulative<u64>,
+}
+
+// The name of the kstat module containing virtual machine kstats.
+const VMM_KSTAT_MODULE_NAME: &str = "vmm";
+
+// The name of the kstat with virtual machine metadata (VM name currently).
+const VM_KSTAT_NAME: &str = "vm";
+
+// The named kstat holding the virtual machine's name. This is currently the
+// UUID assigned by the control plane to the virtual machine instance.
+const VM_NAME_KSTAT: &str = "vm_name";
+
+// The name of kstat containing vCPU usage data.
+const VCPU_KSTAT_PREFIX: &str = "vcpu";
+
+// Prefix for all named data with a valid vCPU microstate that we track.
+const VCPU_MICROSTATE_PREFIX: &str = "time_";
+
+// The number of expected vCPU microstates we track. This isn't load-bearing,
+// and only used to help preallocate an array holding the `VcpuUsage` samples.
+const N_VCPU_MICROSTATES: usize = 6;
+
+#[cfg(not(test))]
+impl KstatTarget for VirtualMachine {
+    // The VMM kstats are organized like so:
+    //
+    // - module: vmm
+    // - instance: a kernel-assigned integer
+    // - name: vm -> generic VM info, vcpuX -> info for each vCPU
+    //
+    // At this part of the code, we don't have that kstat instance, only the
+    // virtual machine instance's control plane UUID. However, the VM's "name"
+    // is assigned to be that control plane UUID in the hypervisor. See
+    // https://github.com/oxidecomputer/propolis/blob/759bf4a19990404c135e608afbe0d38b70bfa370/bin/propolis-server/src/lib/vm/mod.rs#L420
+    // for the current code which does that.
+    //
+    // That means we need to indicate interest in both the `vm` and `vcpuX`
+    // kstats for any instance, and then filter to the right instance in the
+    // `to_samples()` method below, because interest is defined on each
```

Comment:
This really feels like something which propolis should be minding itself, rather than pushing that responsibility to oximeter and/or sled-agent.  It's propolis which chooses the naming for the kernel vmm resource, and should probably query the kstats.

Furthermore, I'm not sure how much of the kstat structure we want to turn into a Committed interface in the data.  Do we want to decide on a more constrained model for how vCPUs can spend their time, rather than directly encoding the kstat names (`time_emu_kern`, etc) into the data?

Reply:
And to be clear about the mapping from the actual kstat run states to those three proposed, I think we'd have:

- `run == time_run + time_emu_kern + time_emu_user`
- `idle == time_init + time_idle`
- `waiting == time_sched`

Reply:
> Propolis _is_ actually querying the kstats, even though the functionality for converting from that to oximeter `Sample`s lives here.

If propolis is otherwise calling into an abstraction provided by oximeter to do that querying, I'm not sure how much that counts as doing the querying.  Propolis has tight coupling to the kernel vmm interfaces (out of necessity), with explicit logic for handling when that interface undergoes change.

> I do think there's value in the `VirtualMachine` definition living in a public, shared place. I'm trying to avoid someone having a conflicting definition, or even a duplicate definition.

Agreed.

> If that's not desirable, I think it's fine to move the target / metric definitions and the `impl KstatTarget` both to propolis. It does mean that anyone wishing to publish oximeter stats about the VM would need to either depend on Propolis or copy that `Target` implementation. I honestly don't feel that strongly, since neither seems without its own drawbacks.

I think something like that would be the right plan.  I'm not sure where, other than propolis, we would expect VM-related oximeter metrics to come from.

> If you were debugging unexpected behavior in Propolis, would you like to be able to see each of these states historically?

If I'm debugging propolis, then sure, but I'm not sure that detail is worth the trade-off of committing to those states as the interface.

> Maybe three states: _doing work_, _not doing work_, and _not able to do work_ (waiting for some reason)? That would be something like `run`, `idle`, and `waiting`.

Yeah, something like what you described sounds right.  Perhaps we might want to split out time spent in emulation (not distinguishing to the consumer whether it's userland or kernel).

Reply:
Sweet, that all sounds good to me. To recap:

- Move the `VirtualMachine` and `VcpuUsage` types to Propolis
- Move `impl KstatTarget for VirtualMachine` to Propolis too
- Specify states as: `run`, `emulation`, `idle`, and `blocked`

If that sounds fine, I'll get it moving this afternoon.